sensorprocess/sensorprocess.go

```diff
@@ -12,39 +11,85 @@ import (
 	s "github.com/viamrobotics/viam-cartographer/sensors"
 )
 
-// defaultTime is used to check if timestamps have not been set yet.
-var defaultTime = time.Time{}
-
 // Config holds config needed throughout the process of adding a sensor reading to the cartofacade.
 type Config struct {
-	CartoFacade            cartofacade.Interface
-	sensorProcessStartTime time.Time
-	IsOnline               bool
-
-	Lidar            s.TimedLidar
-	currentLidarData *s.TimedLidarReadingResponse
+	CartoFacade cartofacade.Interface
+	IsOnline    bool
 
-	IMU            s.TimedMovementSensor
-	currentIMUData *s.TimedIMUReadingResponse
+	Lidar s.TimedLidar
+	IMU   s.TimedMovementSensor
 
 	Timeout                  time.Duration
 	InternalTimeout          time.Duration
 	Logger                   logging.Logger
 	RunFinalOptimizationFunc func(context.Context, time.Duration) error
-
-	Mutex *sync.Mutex
 }
 
-// Update currentIMUData under a mutex lock.
-func (config *Config) updateMutexProtectedIMUData(data s.TimedIMUReadingResponse) {
-	config.Mutex.Lock()
-	config.currentIMUData = &data
-	config.Mutex.Unlock()
+// StartOfflineSensorProcess starts the process of adding lidar and movement sensor data
+// in a deterministically defined order to cartographer.
+func (config *Config) StartOfflineSensorProcess(ctx context.Context) bool {
+	// get the initial lidar reading
+	lidarReading, lidarEndOfDataSetReached, err := getTimedLidarReading(ctx, config)
+	if err != nil || lidarEndOfDataSetReached {
+		return lidarEndOfDataSetReached
+	}
+
+	var imuReading s.TimedIMUReadingResponse
+	if config.IMU != nil {
+		// get the initial IMU reading; discard all IMU readings that were recorded before the first lidar reading
+		for {
```

Comment:
[opt] For ease of reading I would split this for loop into a function so the start up of the offline loop is as simple as:
- get initial lidar reading
- if IMU exists, get initial imu reading

Reply:
Will consider this in my next PR when I add the odometer to it.